presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestAddNotNullFiltersToJoinNode.java

```diff
@@ -0,0 +1,321 @@
+                JOINS_NOT_NULL_INFERENCE_STRATEGY, USE_FUNCTION_METADATA.toString()));
+
+        functionAndTypeManager = createTestFunctionAndTypeManager();
+        Metadata metadata = MetadataManager.createTestMetadataManager();
+        rowExpressionTranslator = new TestingRowExpressionTranslator(metadata);
+    }
+
+    @DataProvider
+    public static Object[][] getExistingNotNullVarsTestCases()
+    {
+        return new Object[][] {
+                {"a IS NOT NULL AND b IS NOT NULL", new String[] {"a", "b"}},
+                {"a > 10 AND b IS NOT NULL AND c is NOT NULL", new String[] {"b", "c"}},
+                {"a is NULL AND b IS NOT NULL", new String[] {"b"}},
+                {"NOT(a is NULL)", new String[] {"a"}},
+                {"NOT(a is NULL OR b is NULL)", new String[] {}},
+                {"a is NOT NULL OR b is NOT NULL", new String[] {}}
+        };
+    }
+
+    @DataProvider
+    public static Object[][] nonStandardOperatorTestCases()
+    {
+        return new Object[][] {
+                {"NOT (b + 10 > c)", new String[] {"b", "c"}},
+                {"abs(b + c) > 10", new String[] {"b", "c"}},
+                {"random(b) = ceil(c)", new String[] {"b", "c"}},
+        };
+    }
+
+    @DataProvider
+    public static Object[][] standardOperatorTestCases()
+    {
+        return new Object[][] {
+                {"a + b > 10", new String[] {"a", "b"}},
+                {"a != 10 - b", new String[] {"a", "b"}},
+                {"a > b", new String[] {"a", "b"}},
+                {"a + NULL > b", new String[] {"a", "b"}},
+                // We can infer NOT NULL predicates on arguments of an AND expression
+                {"a > b and c = d", new String[] {"a", "b", "c", "d"}},
+                {"a IS NULL and b > c", new String[] {"b", "c"}},
+                // We cannot infer NOT NULL predicates on arguments of an OR expression
+                {"a > b OR c = d", new String[] {}},
+                // COALESCE can operate on NULL arguments, so cant infer predicates on its arguments
+                {"COALESCE(a,b)", new String[] {}},
+                // IN can operate on NULL arguments, so cant infer predicates on its arguments
+                {"a IN (b,10,NULL)", new String[] {}},
+                // arr[3] = 10 translates to EQUAL(SUBSCRIPT(arr, 3), 10). SUBSCRIPT is a standard Operator, so we can infer that 'arr' is NOT NULL
+                {"arr[3] = 10", new String[] {"arr"}},
+                // c_struct.a = 10 translates to EQUAL(DEREFERENCE(c_struct, 0), 10).
+                // We chose to not make any inferences for DEREFERENCE clauses, hence we don't add any NOT NULL clauses for 'c_struct' or 'c_struct.a'
+                {"c_struct.a = 10", new String[] {}},
+        };
+    }
+
+    @Test
+    public void testNotNullPredicatesAddedForSingleEquiJoinClause()
+    {
+        String query = "select 1 from lineitem l join orders o on l.orderkey = o.orderkey";
```

Comment:
for all these queries in the plan test, it's a good idea to add them as correctness tests too in AbstractTestJoinQueries